README.md
# Keywords from the UniProt database

UniProt database provides a list of controlled vocabulary represented as keywords
for genes or proteins (https://www.uniprot.org/keywords/). This is useful for summarizing gene functions in a compact way. This package
provides data of keywords hierarchy and gene-keywords relations.


### Install

If you want the latest version, install it directly from GitHub:

```r
if (!requireNamespace("devtools", quietly = TRUE))
    install.packages("devtools")
devtools::install_github("jokergoo/UniProtKeywords")
```

### Usage


First load the package:

```r
library(UniProtKeywords)
```

The package has five data objects. The first contains basic information of every keyword term:

```r
data(kw_terms)
kw_terms[["Cell cycle"]]
```


The other four contain hierarchical structures of the keyword terms:

```r
data(kw_parents)
kw_parents[1:2]

data(kw_children)
kw_children[1:2]

data(kw_ancestors)
kw_ancestors[1:2]

data(kw_offspring)
kw_offspring[1:2]
```

The **UniProtKeywords** package has also compiled genesets of keywords for some species, which can get by the function `load_keyword_genesets()`.
The argument is the taxon ID of a species:

```r
gl = load_keyword_genesets(9606)
gl[1:2]
```

### License

MIT @ Zuguang Gu




NEWS
CHANGES in VERSION 0.99.7

- update to version: 2023_04 of 13-Sep-2023

--------------------------

CHANGES in VERSION 0.99.3

- add the object `UniProtKeywords` which shows the release and source information of the data

--------------------------


tests/test-all.R
library(testthat)
suppressWarnings(suppressPackageStartupMessages(library(UniProtKeywords)))

test_check("UniProtKeywords")


tests/testthat/test-general.R

test_that("load_keyword_genesets", {
	load_keyword_genesets(9606)
	load_keyword_genesets(10090)
	load_keyword_genesets(category = "Domain")

	expect_error(load_keyword_genesets(12345), "not supported")
	expect_error(load_keyword_genesets(category = "aaa"), "not allowed")
})

R/data.R

# == title (variable:UniProtKeywords)
# Release and source information
#
# == example
# UniProtKeywords
UniProtKeywords = list(
	release = "2023_04 of 13-Sep-2023",
	built_date = "2023-09-23",
	source = "https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/docs/keywlist.txt"
)
class(UniProtKeywords) = "UniProtKeywords_info"

# == title
# Print the UniProtKeywords object
#
# == param
# -x A ``UniProtKeyword`` object.
# -... Other arguments
#
# == example
# UniProtKeywords
print.UniProtKeywords_info = function(x, ...) {
	cat("UniProt Keywords\n")
	cat("  Release:", x$release, "\n")
	cat("  Source:", x$source, "\n")
	cat("  Number of keywords:", length(kw_terms), "\n")
	cat("  Built date:", x$built_date, "\n")
}

# == title (data:kw_terms)
# Keyword terms
#
# == details
# Data is from https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/docs/keywlist.txt
#
# Last updated: 2023-09-23.
#
# == value
# A list of keyword terms. Each keyword term has the following elements:
#
# - Identifier
# - Accession
# - Description
# - Synonyms
# - Gene_ontology
# - Hierarchy
# - WWW_site
# - Category
#
# == example
# data(kw_terms)
# kw_terms[[1]]
#



# == title (data:kw_parents)
# Parents keyword terms
#
# == value
# A list of parent keywords.
#
# == example
# data(kw_parents)
# kw_parents[1:2]
#

# == title (data:kw_children)
# child keyword terms
#
# == value
# A list of child keywords.
#
# == example
# data(kw_children)
# kw_children[1:2]
#

# == title (data:kw_ancestors)
# Ancestor keyword terms
#
# == value
# A list of ancestor keywords.
#
# == example
# data(kw_ancestors)
# kw_ancestors[1:2]

# == title (data:kw_offspring)
# Offspring keyword terms
#
# == value
# A list of offspring keywords.
#
# == example
# data(kw_offspring)
# kw_offspring[1:2]



inst/extdata/process/extract.pl
use strict;

## https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.dat.gz

open FILE, "gunzip -c uniprot_sprot.dat.gz |";

my $id;
my $taxon_id;
my $organism;
my $gene_id;
my $keyword;

open OUT, ">uniprot_keywords.txt";
print OUT "Uniprot_ID\tTaxon_ID\tEntrez_Gene_ID\tKeywords\n";

while(my $line = <FILE>) {
	if($line =~/^ID\s/) {

		if($line =~/^ID\s+(\w+)\s/) {
			$id = $1;
		}

		$taxon_id = "";
		$organism = "";
		$gene_id = "";
		$keyword = "";
			
		while($line = <FILE>) {

			chomp $line;

			if($line =~/^\/\//) {
				if($keyword ne "" and $gene_id ne "" and $taxon_id ne "") {
					print OUT "$id\t$taxon_id\t$gene_id\t$keyword\n";
				}
				last;
			}

			if($line =~/^OX\s+NCBI_TaxID=(\d+);$/) {
				$taxon_id = $1;
			}

			if($line =~/^OS\s+(\S.*)$/) {
				$organism = $1;
				while(<FILE> =~/^OS\s+(\S.*)$/) {
					$organism .= " ".$1;
				}
			}

			if($line =~/^DR\s+GeneID; (\d+);/) {
				$gene_id = $1;
			}

			if($line =~/^KW\s+(\w.*)$/) {
				$keyword = $1;
				while(<FILE> =~/^KW\s+(\w.*)$/) {
					$keyword .= " ".$1;
				}
			}
		}

	}
}


inst/extdata/process/process_keywords_ontology.R

library(GetoptLong)

options("timeout" = 99999999)

## https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/docs/keywlist.txt

ln = readLines(url("https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/docs/keywlist.txt"))
i1 = grep("_____________", ln)
i2 = grep("^//", ln)

ln = ln[seq(min(i1)+1, max(i2))]


parse_kw = function(x) {
	x2 = strsplit(x, "   ")
	lt = list()

	CA = NULL
	for(i in seq_along(x2)) {
		code = x2[[i]][1]

		if(code == "CA") next

		if(code == "HI") {
			CA = c(CA, gsub("^(\\w.*?):.*$", "\\1", x2[[i]][2]))
		}

		if(is.null(lt[[code]])) {
			lt[[code]] = x2[[i]][2]
		} else {
			if(code == "DE") {
				lt[[code]] = paste0(lt[[code]], " ", x2[[i]][2])
			} else {
				lt[[code]] = c(lt[[code]], x2[[i]][2])
			}
		}
	}
	if(!is.null(CA)) {
		lt$CA = unique(CA)
	}
	for(nm in names(lt)) {
		if(!nm %in% c("DE")) {
			lt[[nm]] = gsub("(;|\\.)$", "", lt[[nm]])
		}
	}
	if("IC" %in% names(lt)) {
		names(lt)[names(lt) == "IC"] = "ID"
		lt$ID = gsub("\\.$", "", lt$ID)
	}

	name_map = c("ID" = "Identifier",
		         "AC" = "Accession",
		         "DE" = "Description",
		         "SY" = "Synonyms",
		         "GO" = "Gene_ontology",
		         "HI" = "Hierarchy",
		         "WW" = "WWW_site",
		         "CA" = "Category")
	names(lt) = name_map[names(lt)]
	lt
}


k = which(ln == "//")
k = c(0, k)

kw_terms = NULL
for(i in seq_along(k)[-length(k)]) {
	x = ln[seq(k[i]+1, k[i+1] - 1)]

	kw_terms[[i]] = parse_kw(x)
}

names(kw_terms) = sapply(kw_terms, function(x) x$Identifier)


extract_parents = function(hi) {
	if(is.null(hi)) {
		return(NULL)
	}

	x = gsub(":", ";", hi)
	x2 = lapply(strsplit(x, "; "), function(v) {
		v[seq(length(v)-1, length(v))]
	})

	unique(do.call(rbind, x2))
}


keyword_relations = do.call(rbind, lapply(kw_terms, function(x) extract_parents(x$Hierarchy)))
colnames(keyword_relations) = c("parent", "child")

kw_parents = split(keyword_relations[, 1], keyword_relations[, 2])
kw_children = split(keyword_relations[, 2], keyword_relations[, 1])


library(igraph)

g = graph.edgelist(keyword_relations)

# upstream
cm = distances(g, mode = "in")
kw_ancestors = lapply(1:nrow(cm), function(i) {
	x = cm[i, ]
	l = is.finite(x)
	V(g)$name[setdiff(which(l), i)]
})
names(kw_ancestors) = V(g)$name
kw_ancestors = kw_ancestors[sapply(kw_ancestors, length) > 0]


# downstream
cm = distances(g, mode = "out")
kw_offspring = lapply(1:nrow(cm), function(i) {
	x = cm[i, ]
	l = is.finite(x)
	V(g)$name[setdiff(which(l), i)]
})
names(kw_offspring) = V(g)$name
kw_offspring = kw_offspring[sapply(kw_offspring, length) > 0]


save(kw_parents, file = "../../../data/kw_parents.rda", compress = "xz")
save(kw_children, file = "../../../data/kw_children.rda", compress = "xz")
save(kw_ancestors, file = "../../../data/kw_ancestors.rda", compress = "xz")
save(kw_offspring, file = "../../../data/kw_offspring.rda", compress = "xz")
save(kw_terms, file = "../../../data/kw_terms.rda", compress = "xz")

### annotation to genes
## uniprot_keywords.txt is generated by extract.pl

if(file.exists("uniprot_sprot.dat.gz")) file.remove("uniprot_sprot.dat.gz")
download.file("https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.dat.gz", dest = "uniprot_sprot.dat.gz")
system("perl extract.pl")

tb = read.table("uniprot_keywords.txt", sep = "\t", header = T)

format_geneset = function(tb2) {
	g = tb2$Entrez_Gene_ID
	g = as.character(g)
	keywords = tb2$Keywords
	keywords = gsub("\\.$", "", keywords)
	keywords = strsplit(keywords, "; ")

	g = rep(g, times = sapply(keywords, length))
	keywords = unlist(keywords)

	gl = split(g, keywords)

	for(nm in names(gl)) {
		offspring = kw_offspring[[nm]]
		if(!is.null(offspring)) {
			gl[[nm]] = unique(unlist(gl[ offspring ]))
		}
	}

	gl

}


for(t in unique(tb$Taxon_ID)) {
	l = tb$Taxon_ID == t
	if(sum(l) >= 1000) {
		tb2 = tb[l, ]
		lt = format_geneset(tb2)
		saveRDS(lt, file = qq("../keywords_anno/keywords_@{t}_gene_annotation.rds"), compress = "xz")
	}
}


# if(file.exists("uniprot_sprot.dat.gz")) file.remove("uniprot_sprot.dat.gz")






R/load.R

# == title
# Load keyword genesets for a specific species
#
# == param
# -taxon_id The taxon ID. To make it more flexible, you can also provide the Latin name or the normal name of the species.
# -category Category of keywords. There are the following categories: "Biological process", "Cellular component", "Coding sequence diversity", "Developmental stage", "Disease", "Domain", "Ligand", "Molecular function", "Post-translational modification", "Technical term". 
# -as_table If true, the returned value will be a two-column data frame.
#
# == details
# Following are the supported species (with more than 1000 genes annotated):
#
# - "10090": Mus musculus / house mouse
# - "10116": Rattus norvegicus / Norway rat
# - "208964": Pseudomonas aeruginosa PAO1 / strain, g-proteobacteria
# - "224308": Bacillus subtilis subsp. subtilis str. 168 / strain, firmicutes
# - "237561": Candida albicans SC5314 / strain, budding yeasts
# - "243232": Methanocaldococcus jannaschii DSM 2661 / strain, euryarchaeotes
# - "284812": Schizosaccharomyces pombe 972h- / strain, ascomycete fungi
# - "3702": Arabidopsis thaliana / thale cress
# - "39947": Oryza sativa Japonica Group / (Japanese rice), monocots
# - "44689": Dictyostelium discoideum / species, cellular slime molds
# - "559292": Saccharomyces cerevisiae S288C / strain, budding yeasts
# - "6239": Caenorhabditis elegans / species, nematodes
# - "623": Shigella flexneri / species, enterobacteria
# - "7227": Drosophila melanogaster / (fruit fly), species, flies
# - "7955": Danio rerio / (zebrafish), species, bony fishes
# - "83332": Mycobacterium tuberculosis H37Rv / strain, high G+C Gram-positive bacteria
# - "83333": Escherichia coli K-12 / strain, enterobacteria
# - "83334": Escherichia coli O157:H7 / serotype, enterobacteria
# - "8355": Xenopus laevis / (African clawed frog), species, frogs & toads
# - "8364": Xenopus tropicalis / (tropical clawed frog), species, frogs & toads
# - "9031": Gallus gallus / (chicken), species, birds
# - "9601": Pongo abelii / (Sumatran orangutan), species, primates
# - "9606": Homo sapiens / human
# - "9823": Sus scrofa / (pig), species, even-toed ungulates
# - "9913": Bos taurus / cattle
# - "99287": Salmonella enterica subsp. enterica serovar Typhimurium str. LT2 / strain, enterobacteria
#
# == value
# If ``as_table`` is set to ``FALSE``, it returns a list of gene sets where Entrez IDs are the gene IDs.
# If ``as_table`` is set to ``TRUE``, it returns a two-column data frame.
#
# == example
# lt = load_keyword_genesets(9606)
# lt[3:4]
# tb = load_keyword_genesets(9606, as_table = TRUE)
# head(tb)
# # load_keyword_genesets("mouse")
load_keyword_genesets = function(taxon_id = 9606, category = NULL, as_table = FALSE) {

	if(!grepl("^\\d+$", taxon_id)) {
		i = grep(taxon_id, ORGANISM[, 2], ignore.case = TRUE)
		if(length(i) == 1) {
			taxon_id = ORGANISM[i, 1]
		} else {
			i = grep(taxon_id, ORGANISM[, 3], ignore.case = TRUE)
			if(length(i) == 1) {
				taxon_id = ORGANISM[i, 1]
			} else {
				stop("Species '", taxon_id, "' is not supported.")
			}
		}
	}

	f = system.file("extdata", "keywords_anno", paste0("keywords_", taxon_id, "_gene_annotation.rds"), package = "UniProtKeywords")
	if(file.exists(f)) {
		lt = readRDS(f)
	} else {
		stop("Species '", taxon_id, "' is not supported.")
	}

	if(!is.null(category)) {

		if(any(category == "Post-translational modification")) {
			category[category == "Post-translational modification"] = "PTM"
		}

		diff = setdiff(category, ALL_KEYWORDS_CATEGORIES)
		if(length(setdiff(category, ALL_KEYWORDS_CATEGORIES)) == 1) {
			stop("Following value for `category` is not allowed: ", diff)
		} else if(length(setdiff(category, ALL_KEYWORDS_CATEGORIES)) > 1) {
			not_allowed = paste(diff, collapse = ", ")
			stop("Following values for `category` are not allowed: ", not_allowed)
		}

		l = vapply(names(lt), function(nm) {
			any(kw_terms[[nm]]$Category %in% category)
		}, TRUE)
		lt = lt[l]
	}
	lt = lapply(lt, as.character)

	if(as_table) {
		tb = data.frame(keyword = rep(names(lt), times = vapply(lt, length, 0)),
			            gene = unlist(lt))
		rownames(tb) = NULL
		return(tb)
	} else {

		return(lt)
	}
}

ALL_KEYWORDS_CATEGORIES = c("Biological process", "Cellular component", "Coding sequence diversity", 
	                        "Developmental stage", "Disease", "Domain", "Ligand", "Molecular function", 
	                        "Post-translational modification", "PTM", "Technical term")


.onLoad <- function(lib, pkg) {
	data(kw_terms, package = pkg, envir = topenv())
}


ORGANISM = read.table(textConnection(
"10090;Mus musculus;house mouse
10116;Rattus norvegicus;Norway rat
208964;Pseudomonas aeruginosa PAO1;strain, g-proteobacteria
224308;Bacillus subtilis subsp. subtilis str. 168;strain, firmicutes
237561;Candida albicans SC5314;strain, budding yeasts
243232;Methanocaldococcus jannaschii DSM 2661;strain, euryarchaeotes
284812;Schizosaccharomyces pombe 972h-;strain, ascomycete fungi
3702;Arabidopsis thaliana;thale cress
39947;Oryza sativa Japonica Group;(Japanese rice), monocots
44689;Dictyostelium discoideum;species, cellular slime molds
559292;Saccharomyces cerevisiae S288C;strain, budding yeasts
6239;Caenorhabditis elegans;species, nematodes
623;Shigella flexneri;species, enterobacteria
7227;Drosophila melanogaster;(fruit fly), species, flies
7955;Danio rerio;(zebrafish), species, bony fishes
83332;Mycobacterium tuberculosis H37Rv;strain, high G+C Gram-positive bacteria
3333;Escherichia coli K-12;strain, enterobacteria
83334;Escherichia coli O157:H7;serotype, enterobacteria
8355;Xenopus laevis;(African clawed frog), species, frogs & toads
8364;Xenopus tropicalis;(tropical clawed frog), species, frogs & toads
9031;Gallus gallus;(chicken), species, birds
9601;Pongo abelii;(Sumatran orangutan), species, primates
9606;Homo sapiens;human
9823;Sus scrofa;(pig), species, even-toed ungulates
9913;Bos taurus;cattle
99287;Salmonella enterica subsp. enterica serovar Typhimurium str. LT2;strain, enterobacteria
"), sep = ";")

